zebra/dplane_fpm_pb.c

```diff
@@ -0,0 +1,973 @@
+	return prefix;
+}
+
+static ssize_t protobuf_msg_encode(struct zebra_dplane_ctx *ctx, uint8_t *data,
+				   size_t datalen)
+{
+	Fpm__Message *msg;
+	QPB_DECLARE_STACK_ALLOCATOR(allocator, 4096);
+	size_t len;
+
+	QPB_INIT_STACK_ALLOCATOR(allocator);
+
+	msg = create_fpm_message(&allocator, ctx);
+	if (!msg)
+		return 0;
+	len = fpm__message__pack(msg, data);
+	/* not enough space */
+	if (len > datalen)
+		return 0;
+	QPB_RESET_STACK_ALLOCATOR(allocator);
+	return len;
+}
+
+static int fpm_pb_enqueue(struct fpm_pb_ctx *fpc, struct zebra_dplane_ctx *ctx)
+{
+	uint8_t pb_buf[PB_PKT_BUF_SIZE];
+	size_t pb_buf_len;
+	uint64_t obytes, obytes_peak;
+
+	frr_mutex_lock_autounlock(&fpc->obuf_mutex);
+
+	pb_buf_len = protobuf_msg_encode(ctx, pb_buf, sizeof(pb_buf));
+
+	if (pb_buf_len == 0) {
+		/* protobuf msg encode error */
+		return 0;
+	}
+
+
+	/* We must know if someday a message goes beyond 65KiB. */
+	assert((pb_buf_len + FPM_HEADER_SIZE) <= UINT16_MAX);
+
+	/* Check if we have enough buffer space. */
+	if (STREAM_WRITEABLE(fpc->obuf) < (pb_buf_len + FPM_HEADER_SIZE)) {
+		atomic_fetch_add_explicit(&fpc->counters.buffer_full, 1,
+					  memory_order_relaxed);
+
+		if (IS_ZEBRA_DEBUG_FPM)
+			zlog_debug("%s: buffer full: wants to write %zu but has %zu",
+				   __func__, pb_buf_len + FPM_HEADER_SIZE,
+				   STREAM_WRITEABLE(fpc->obuf));
+
+		return -1;
+	}
+
+	/*
+	 * Fill in the FPM header information.
+	 *
+	 * See FPM_HEADER_SIZE definition for more information.
+	 */
+	stream_putc(fpc->obuf, 1);
+	stream_putc(fpc->obuf, 2);
+	stream_putw(fpc->obuf, pb_buf_len + FPM_HEADER_SIZE);
+
+	/* Write current data. */
+	stream_write(fpc->obuf, pb_buf, (size_t)pb_buf_len);
+
+	/* Account number of bytes waiting to be written. */
+	atomic_fetch_add_explicit(&fpc->counters.obuf_bytes,
+				  pb_buf_len + FPM_HEADER_SIZE,
+				  memory_order_relaxed);
+	obytes = atomic_load_explicit(&fpc->counters.obuf_bytes,
+				      memory_order_relaxed);
+	obytes_peak = atomic_load_explicit(&fpc->counters.obuf_peak,
+					   memory_order_relaxed);
+	if (obytes_peak < obytes)
+		atomic_store_explicit(&fpc->counters.obuf_peak, obytes,
+				      memory_order_relaxed);
+
+	/* Tell the thread to start writing. */
+	event_add_write(fpc->fthread->master, fpm_write, fpc, fpc->socket,
+			&fpc->t_write);
+
+	return 0;
+}
+
+static int fpm_pb_start(struct zebra_dplane_provider *prov)
+{
+	struct fpm_pb_ctx *fpc;
+
+	fpc = dplane_provider_get_data(prov);
+	fpc->fthread = frr_pthread_new(NULL, prov_name, prov_name);
+	assert(frr_pthread_run(fpc->fthread, NULL) == 0);
+	fpc->ibuf = stream_new(PB_PKT_BUF_SIZE);
+	fpc->obuf = stream_new(PB_PKT_BUF_SIZE * 128);
+	pthread_mutex_init(&fpc->obuf_mutex, NULL);
+	fpc->socket = -1;
+	fpc->disabled = true;
+	fpc->prov = prov;
+	dplane_ctx_q_init(&fpc->ctxqueue);
+	pthread_mutex_init(&fpc->ctxqueue_mutex, NULL);
+
+	struct sockaddr_in *sin;
+	uint8_t naddr[INET6_BUFSIZ];
+
+	if (inet_pton(AF_INET, "127.0.0.1", naddr) != 1) {
```

Comment:
don't see the value in this conversion - you know the four octets you want to use?

Reply:
Done, we removed the redundant conversion.